pkg/service/admission/endpoint_admission.go

```diff
@@ -0,0 +1,115 @@
+package admission
+
+import (
+	"fmt"
+	"io"
+	"net"
+
+	authorizationapi "github.com/openshift/origin/pkg/authorization/api"
+	"github.com/openshift/origin/pkg/client"
+	oadmission "github.com/openshift/origin/pkg/cmd/server/admission"
+
+	kadmission "k8s.io/kubernetes/pkg/admission"
+	kapi "k8s.io/kubernetes/pkg/api"
+	clientset "k8s.io/kubernetes/pkg/client/clientset_generated/internalclientset"
+)
+
+const RestrictedEndpointsPluginName = "RestrictedEndpointsAdmission"
+
+func init() {
+	kadmission.RegisterPlugin("RestrictedEndpointsAdmission", func(client clientset.Interface, config io.Reader) (kadmission.Interface, error) {
+		return NewRestrictedEndpointsAdmission(nil), nil
+	})
+}
+
+type restrictedEndpointsAdmission struct {
+	*kadmission.Handler
+
+	client             client.Interface
+	restrictedNetworks []*net.IPNet
+}
+
+var _ = oadmission.WantsOpenshiftClient(&restrictedEndpointsAdmission{})
+
+// ParseSimpleCIDRRules parses a list of CIDR strings
+func ParseSimpleCIDRRules(rules []string) (networks []*net.IPNet, err error) {
+	for _, s := range rules {
+		_, cidr, err := net.ParseCIDR(s)
+		if err != nil {
+			return nil, err
+		}
+		networks = append(networks, cidr)
+	}
+	return networks, nil
+}
+
+// NewRestrictedEndpointsAdmission creates a new endpoints admission plugin.
+func NewRestrictedEndpointsAdmission(restrictedNetworks []*net.IPNet) *restrictedEndpointsAdmission {
+	return &restrictedEndpointsAdmission{
+		Handler:            kadmission.NewHandler(kadmission.Create, kadmission.Update),
+		restrictedNetworks: restrictedNetworks,
+	}
+}
+
+func (r *restrictedEndpointsAdmission) SetOpenshiftClient(c client.Interface) {
+	r.client = c
+}
+
+func (r *restrictedEndpointsAdmission) findRestrictedIP(ep *kapi.Endpoints) string {
+	for _, subset := range ep.Subsets {
+		for _, addr := range subset.Addresses {
+			ip := net.ParseIP(addr.IP)
+			if ip == nil {
+				continue
+			}
+			for _, net := range r.restrictedNetworks {
+				if net.Contains(ip) {
+					return addr.IP
+				}
+			}
+		}
+	}
+	return ""
+}
+
+func resourceName(objectMeta kapi.ObjectMeta) string {
+	if len(objectMeta.GenerateName) > 0 {
+		return objectMeta.GenerateName
+	}
+	return objectMeta.Name
+}
+
+// Admit determines if the endpoints object should be admitted
+func (r *restrictedEndpointsAdmission) Admit(a kadmission.Attributes) error {
+	if a.GetResource().GroupResource() != kapi.Resource("endpoints") {
+		return nil
+	}
+	ep, ok := a.GetObject().(*kapi.Endpoints)
+	if !ok {
+		return nil
+	}
+	restrictedIP := r.findRestrictedIP(ep)
+	if restrictedIP == "" {
+		return nil
+	}
+
+	resource := kapi.Resource(authorizationapi.RestrictedEndpointsResource)
+	subjectAccessReview := authorizationapi.AddUserToLSAR(a.GetUserInfo(),
```

Comment:
+1, forgot about that

@danwinship, have your admission plugin implement a method like this:
```
func (o *podNodeConstraints) SetAuthorizer(a authorizer.Authorizer) {
	o.authorizer = a
}
```

add a check to make sure you conform to the `WantsAuthorizer` interface:
```
var _ = oadmission.WantsAuthorizer(&restrictedEndpointsAdmission{})
```

then, rather than an API call to do the access check, you can check against the same authorizer the API guards itself with (see podNodeConstraints#checkPodsBindAccess for an example how to do this)